src/wallet/partialTxProposal.ts

```diff
@@ -0,0 +1,319 @@
+   * @param {Object} [options]
+   * @param {number|null} [options.timelock=null] UNIX timestamp of the timelock.
+   * @param {boolean} [options.isChange=false] If the output should be considered as change.
+   * @param {number} [options.tokenData=0] TokenData of the Output.
+   *
+   * @throws AddressError
+   */
+  addOutput(
+    token: string,
+    value: number,
+    address: string,
+    {
+      timelock = null,
+      isChange = false,
+      tokenData = 0
+    }: { timelock?: number|null, isChange?: boolean, tokenData?: number } = {}
+  ) {
+    this.resetSignatures();
+
+    const addr = new Address(address, {network: this.network});
+    let script;
+    switch(addr.getType()) {
+      case OutputType.P2SH:
+        script = new P2SH(addr, { timelock });
+        break
+      case OutputType.P2PKH:
+        script = new P2PKH(addr, { timelock });
+        break
+      default:
+        throw new AddressError('Unsupported address type');
+    }
+    this.partialTx.addOutput(value, script.createScript(), tokenData, { token, isChange });
+  }
+
+  /**
+   * Reset any data calculated from the partial tx.
+   */
+  resetSignatures() {
+    this.signatures = null;
+    this.transaction = null;
+  }
+
+  /**
+   * Unmark all inputs currently on the partial tx as not `selected_as_input`.
+   */
+  unmarkAsSelected() {
+    const walletData = wallet.getWalletData();
+    const historyTransactions = walletData['historyTransactions'] || {};
+
+    for(const input of this.partialTx.inputs) {
+      // The input may not be present on the loaded wallet's history
+      if(input.hash in historyTransactions) {
+        historyTransactions[input.hash].outputs[input.index]['selected_as_input'] = false;
+      }
+    }
+  }
+
+  /**
+   * Returns true if the transaction funds are balanced and the signatures match all inputs.
+   *
+   * @returns {boolean}
+   */
+  isComplete() {
+    return (!!this.signatures) && this.partialTx.isComplete() && this.signatures.isComplete();
+  }
+
+  /**
+   * Create the data to sign from the current transaction signing the loaded wallet inputs.
+   *
+   * @param {string} pin The loaded wallet's pin to sign the transaction.
+   *
+   * @throws {InvalidPartialTxError} Inputs and outputs balance should match before signing.
+   * @throws {UnsupportedScriptError} When we have an unsupported output script.
+   * @throws {IndexOOBError} input index should be inside the inputs array.
+   */
+  async signData(pin: string) {
+    if (!this.partialTx.isComplete()) {
+      // partialTx is not complete, we cannot sign it.
+      throw new InvalidPartialTxError('Cannot sign incomplete data');
+    }
+
+    const tx: Transaction = this.partialTx.getTx();
+
+    this.signatures = new PartialTxInputData(
+      tx.getDataToSign().toString('hex'),
+      tx.inputs.length
+    );
+
+    // The validation method populates the addresses
+    const valid = await this.partialTx.validate();
```

Comment:
This method goes to the full node to validate the tx. From our last conversation I thought we would have a parameter to do it or not, so we could sign the transactions offline. Am I missing something here?